### ISSUE TITLE
Question about the Load balancing loss that was used

### ISSUE THREAD

**Issue by andresnowak** [allenai/OLMoE#45]
Hello,  I wanted to ask a question about the balancing loss that was used in the paper. From what I understand the balancing loss that was used was this one https://github.com/databricks/megablocks/blob/78eea65fda01e638af36ae38853bc51efb04a4b4/megablocks/layers/moe.py#L32 as mentioned in this issue #40. So the formula being used here is $L_{LB} = N_E \cdot \sum_{i=1}^{N_E} f_i \cdot P_i$, where $f_i = \frac{1}{KT} \sum_{x \in T} \text{routing-map}(x, i)$ and $P_i = \frac{1}{T} \sum_{x \in T} probs(x, i)$, where $T = \text{Batch size} \cdot Tokens$ (so the batch wise balancing loss), and the reason I think why in the paper the graph of the balancing loss shows 0.1 is because you visualize $w * L_{BL}$ instead of only $L_{BL}$ (but it should be close to 0.01 not 0.1, because the value being used is 0.01 no?, but maybe in the graph we can't see it). But in the huggingface code of olmoe https://github.com/huggingface/transformers/blob/acb0f22679aa6229d37cec06ae9dbaa7a446bb8e/src/transformers/models/olmoe/modeling_olmoe.py you have this 

$$
\begin{aligned} 
f_i &= \frac{N}{T} \sum_{t=1}^{T} \mathbb{1} (s_{i,t} \in \text{Topk}(\{s_{j,t} \mid 1 \leqslant j \leqslant N\}, K)), \quad \sum^N_i f_i = K \\ 
P_i &= \frac{1}{T} \sum_{t=1}^{T} s_{i,t} 
\end{aligned}
$$

as it seems in the code $f_i$ is never divided by $K$ 

I just wanted to ask this, because I'm trying to understand the different balancing losses (or auxiliary-free balancing) each open-source MoE uses, and just wanted to make sure I know which one was used by olmoe as I got confused with the huggingface code.

**Comment by Muennighoff** [comment]
Yeah this loss is used: https://github.com/Muennighoff/megablocks/blob/4a25bc7b5665bcb9da93d72d5ad0c14d41e1a351/megablocks/layers/moe.py#L31 which should be the one you linked.

The plotting code for that figure is here: https://colab.research.google.com/drive/15PTwmoxcbrwWKG6ErY44hlJlLLKAj7Hx?usp=sharing which visualizes it as it is logged here https://github.com/allenai/OLMo/blob/04a2da53db172bd9a0450705592ed50888bdcaa7/olmo/train.py#L966 which yes is after weighting as the weight is applied before here https://github.com/Muennighoff/megablocks/blob/4a25bc7b5665bcb9da93d72d5ad0c14d41e1a351/megablocks/layers/moe.py#L95

It is 0.1 as you can see in https://wandb.ai/ai2-llm/olmoe/reports/Plot-LBL-vs-No-LBL--Vmlldzo4OTkyNDg4 as for that particular run the weight was 0.1 as you can see in https://wandb.ai/ai2-llm/olmoe/runs/xzq6s5u1/overview

I'm not sure about the HF one - it may have a slightly different LBL normalization as we likely took it from other MoEs that were already there, but you can probably just check the code carefully and compare as you seem to have done. Hope this helps :)

**Comment by andresnowak** [comment]
Thank you very much for the answer and for the quick reply.

**Comment by andresnowak** [comment]
Sorry to bother again, but If I could ask another question, when doing the training, how did you guys choose the Micro-batch size for the training?, because this part also affects the Auxiliary balancing loss no?, as using 1 or 2 MBS would basically almost be the same as doing the sequence-wise one no?, or at least also depending on the sequence length and how many different documents are in that sequence if one is using document attention masking (and at least deepseekv3 showed that batch-wise is better than sequence wise if I understood their paper)?. Just trying to know if you guys also have an insight into this, because I never see this part talked about in the MoE papers (and sorry I don't know how to read the wandb runs, I Just don't know like which one was the final one because I see many runs that only take 10 min, sorry for my ignorance in that part)

**Comment by Muennighoff** [comment]
I don't remember doing anything special for that so yeah if MBS=1 it will seek to balance the load for a single sequence; I just went with the largest MBS that would fit GPU memory to have it run fast. Seems like a good idea to make it batch-level in future MoEs or maybe somehow remove it

**Comment by andresnowak** [comment]
Really you think so?, because at least from DeepseekV3 they found that batch-wise had better performance than sequence-wise no?. And at least seeing what each one does, sequence-wise makes it so that all experts will see the same type of topic domains, and batch-wise (if it's not too big or small) makes it so that at least there is a higher probability that each expert will be able to see different topics no? (For the supposed expert specialization).

And yeah from the conclusion you guys put on the paper, it still seems interesting to me if in reality is this balancing good or bad, maybe we are giving the experts too little credit to also be able to learn what type of tokens they want to attend to even if they are forced to see them  with this uniform balancing.

And thank you for your answer.

**Comment by Muennighoff** [comment]
Oh sorry I meant batch-level (just edited it)! Cuz as you say batch-level gives more flexibility than sequence-level - whatever gives the experts more flexibility is better I think (which aligns with what we wrote in the paper as you say)

**Comment by andresnowak** [comment]
Okay makes sense, and thank you very much for your time and answers.
